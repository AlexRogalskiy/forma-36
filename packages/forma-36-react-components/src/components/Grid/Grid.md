# Grid

[CSS Grid](https://developer.mozilla.org/en-US/docs/Glossary/Grid) based react component, comes with predefined values to ensure design consistency, and ease of use.

```js
import { Grid, GridItem } from '@contenful/forma-36-react-components';
```

## API
### columns
Accepts a number When defined as a number, it will split the space into equally sized columns, otherwise it accepts any of the [grid-template-columns](https://developer.mozilla.org/en-US/docs/Web/CSS/grid-template-columns) css properties. e.g.

```jsx
<Grid columns={6}></Grid>
// Or 
<Grid columns={'auto 1fr'}></Grid>
```


### rows
Accepts a number or any of [grid-template-rows](https://developer.mozilla.org/en-US/docs/Web/CSS/grid-template-rows) css properties.

```jsx
<Grid rows={6}></Grid>
// Or 
<Grid rows={'auto 1fr'}></Grid>
```

### columnGap & rowGap
`columnGap` represents space between columns, `rowGap` represents the space between rows, both accepts one of [spacing](https://f36.contentful.com/foundation/spacing/) token names.

`spacing2xs` | `spacingXs` | `spacingXs` | `spacingS` | `spacingM` | `spacingL` | `spacingXl` | `spacing2Xl` | `spacing3Xl` | `spacing4Xl`

**e.g.**

```jsx
  <Grid columnGap='spacingXs' rowGap='spacingXl'></Grid>
```


