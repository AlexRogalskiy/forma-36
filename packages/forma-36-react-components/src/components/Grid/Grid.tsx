import React from 'react';
import * as CSS from 'csstype';
import tokens from '@contentful/forma-36-tokens';
import cn from 'classnames';

import styles from './Grid.css';

export type SpacingProps = 'spacing2xs' | 'spacingXs' | 'spacingXs' | 'spacingS' | 'spacingM' | 'spacingL' | 'spacingXl' | 'spacing2Xl' | 'spacing3Xl' | 'spacing4Xl';

export interface GridProps {
  /**
   * Class names to be appended to the className prop of the component */
  className?: string;
  /**
   * Child nodes to be rendered in the component */
  children?: React.ReactNode;
  /**
   * An ID used for testing purposes applied as a data attribute (data-test-id) */
  testId?: string;
  /**
   * Defines how many columns, default is `auto` */
  columns?: number | CSS.GridTemplateColumnsProperty<string>;
  /**
   * Defines how many rows, default is `auto` */
  rows?: number | CSS.GridTemplateColumnsProperty<string>;
  /**
   * Spaces between rows, corresponds to of spacing tokens values, default is 0 */
  rowGap?: SpacingProps;
  /**
   * One of Spacing tokens values, default is 0 */
  columnGap?: SpacingProps;
  /**
   * One of Spacing tokens values, default is 0 */
  flow?: CSS.GridAutoFlowProperty
  /**
   * Sets display:inline-grid */
  inline?: boolean;
  /**
   * One of justify-content css values */
  justifyContent?: CSS.JustifyContentProperty;
  /**
   * One of justify-content css values */
  alignContent?: CSS.AlignContentProperty;
  /**
   * style prop, for inline styles */
  style?: React.CSSProperties;
}

const defaultProps = {
  columns: 'auto',
  rows: 'auto',
  columnGap: 'spacingM',
  rowGap: 'none',
  testId: 'cf-ui-grid',
};

export const Grid = (props: GridProps) => {
  const {
    className,
    children,
    testId,
    rows,
    columns,
    inline,
    rowGap,
    columnGap,
    justifyContent,
    alignContent,
    style,
    ...otherProps
  } = props;

  const handleGridTemplate = (value?: string | number) => {
    if(typeof value === 'number') {
      return `repeat(${value}, minmax(0, 1fr)`
    }
    return value;
  }

  const inlineStyle = {
    gridTemplateColumns: handleGridTemplate(columns),
    gridTemplateRows: handleGridTemplate(rows),
    justifyContent,
    alignContent,
    gridColumnGap: columnGap && tokens[columnGap],
    gridRowGap: rowGap && tokens[rowGap],
    ...style
  }

  const classNames = cn(styles.Grid, className, {
    [styles.Grid__inline]: inline
  });

  return (
    <div {...otherProps} style={inlineStyle} className={classNames} data-test-id={testId}>
      {children}
    </div>
  );
}
Grid.defaultProps = defaultProps;

export default Grid;
