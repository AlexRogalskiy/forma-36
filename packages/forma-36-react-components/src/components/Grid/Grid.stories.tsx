import React from 'react';
import { storiesOf } from '@storybook/react';
import { text, number } from '@storybook/addon-knobs';
import tokens from '@contentful/forma-36-tokens';

import Grid from './Grid';
import GridItem from './GridItem';

const DemoBoxes = ({times}: {times: number}) => {
  const styles = {
    height: '100px',
    backgroundColor: tokens.colorIceDark
  }
  let result = [];
  for (let i=0; i<=times; i++) {
    result.push(
      <GridItem style={styles}></GridItem>
    )
  }
  return <>{result}</>
}

storiesOf('Components|Grid', module)
  .addParameters({
    propTypes: Grid['__docgenInfo'],
  })
  .add('default', () => {
    return (
      <Grid 
        columns={number('columns', 6)}
        rows={number('row', 2)}
        rowGap={'spacingS'} 
        className={text('className', '')}>
          <DemoBoxes times={number('Example Items', 17)} />
      </Grid>
    )
  });
