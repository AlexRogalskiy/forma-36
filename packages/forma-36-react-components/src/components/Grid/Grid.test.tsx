import React from 'react';
import { shallow, mount } from 'enzyme';
import { axe } from 'jest-axe';
import Grid from './Grid';

it('renders the component', () => {
  const output = shallow(<Grid>Grid</Grid>);

  expect(output).toMatchSnapshot();
});

it('renders the component with an additional class name', () => {
  const output = shallow(<Grid className="my-extra-class">Grid</Grid>);

  expect(output).toMatchSnapshot();
});

it('has no a11y issues', async () => {
  const output = mount(<Grid>Grid</Grid>).html();
  const results = await axe(output);

  expect(results).toHaveNoViolations();
});
