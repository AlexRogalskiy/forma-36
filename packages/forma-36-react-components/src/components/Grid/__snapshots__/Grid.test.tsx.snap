// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the component 1`] = `
<div
  class="Grid"
  data-test-id="cf-ui-grid"
  style="grid-template-columns: auto; grid-template-rows: auto; column-gap: 1rem; row-gap: 0;"
>
  Grid
</div>
`;

exports[`renders the component with an additional class name 1`] = `
<div
  class="Grid my-extra-class"
  data-test-id="cf-ui-grid"
  style="grid-template-columns: auto; grid-template-rows: auto; column-gap: 1rem; row-gap: 0;"
>
  Grid
</div>
`;

exports[`should have correct styles columnGap should be 1rem by default 1`] = `
<div
  class="Grid"
  data-test-id="cf-ui-grid"
  style="grid-template-columns: auto; grid-template-rows: auto; column-gap: 1rem; row-gap: 0;"
>
  Grid
</div>
`;

exports[`should have correct styles rowGap should be 0 by default 1`] = `
<div
  class="Grid"
  data-test-id="cf-ui-grid"
  style="grid-template-columns: auto; grid-template-rows: auto; column-gap: 1rem; row-gap: 0;"
>
  Grid
</div>
`;

exports[`should have correct styles should has correct numbers of columns 1`] = `
<div
  class="Grid"
  data-test-id="cf-ui-grid"
  style="grid-template-columns: repeat(12, minmax(0, 1fr); grid-template-rows: auto; column-gap: 1rem; row-gap: 0;"
>
  Grid
</div>
`;

exports[`should have correct styles should has correct numbers of rows 1`] = `
<div
  class="Grid"
  data-test-id="cf-ui-grid"
  style="grid-template-columns: auto; grid-template-rows: repeat(12, minmax(0, 1fr); column-gap: 1rem; row-gap: 0;"
>
  Grid
</div>
`;
